Sample Applications/ExpenseIt/ExpenseItDemo/CreateExpenseReportDialogBox.cs

```diff
@@ -82,15 +90,37 @@ private void viewChartButton_Click(object sender, RoutedEventArgs e)
             dlg.Show();
         }
 
+        bool isValid(DependencyObject parent)
+        {
+            if (System.Windows.Controls.Validation.GetHasError(parent)) return false;
+
+            for(int i = 0; i != VisualTreeHelper.GetChildrenCount(parent); i++)
+            {
+                DependencyObject child = VisualTreeHelper.GetChild(parent, i);
+                if (!isValid(child)) return false;
+            }
+
+            return true;
+        }
```

Comment:
Do you really need all that?  I think DataGrid won't leave edit-mode if there are validation errors, so all you have to do is check whether the DataGrid is editing.  (Caution:   I remember arguing about exactly this issue - knowing whether all validation errors have been cleared before proceeding with a higher-level commit like 'save' or 'close dialog' - but I don't remember how it turned out.)